libparsec/crates/client/tests/unit/certificates_ops/add_user_certificate.rs

```diff
@@ -0,0 +1,138 @@
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        AddCertificateError::InvalidCertificate(
+            InvalidCertificateError::ContentAlreadyExists { .. }
+        )
+    )
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn index_already_exists(env: &TestbedEnv) {
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    let store = ops.store.for_write().await;
+    let (_, signed) = env.get_user_certificate("alice");
+
+    ops.add_certificates_batch(&store, 0, [signed.clone()].into_iter())
+        .await
+        .unwrap();
+
+    let err = ops
+        .add_certificates_batch(&store, 0, [signed].into_iter())
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        AddCertificateError::InvalidCertificate(InvalidCertificateError::IndexAlreadyExists { .. })
+    );
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn invalid_timestamp(env: &TestbedEnv) {
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(env, &alice).await;
+
+    let store = ops.store.for_write().await;
+    let (certif, signed) = env.get_user_certificate("alice");
+
+    let mut old_user_certif = (*certif).clone();
+    old_user_certif.timestamp = DateTime::from_ymd_hms_us(1999, 1, 1, 0, 0, 0, 0).unwrap();
+    let old_signed = Bytes::from(old_user_certif.dump_and_sign(&alice.signing_key));
+
+    let err = ops
+        .add_certificates_batch(&store, 0, [signed, old_signed].into_iter())
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        AddCertificateError::InvalidCertificate(InvalidCertificateError::InvalidTimestamp {
+            last_certificate_timestamp,
+            ..
+        })
+        if last_certificate_timestamp == DateTime::from_ymd_hms_us(2000, 1, 2, 0, 0, 0, 0).unwrap()
+    );
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn non_existing_author(env: &TestbedEnv) {
+    let env = env.customize(|builder| {
+        builder.new_user("bob");
+    });
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    let store = ops.store.for_write().await;
+    let (_, signed) = env.get_user_certificate("bob");
+
+    let err = ops
+        .add_certificates_batch(&store, 0, [signed].into_iter())
+        .await
+        .unwrap_err();
+
+    p_assert_matches!(
+        err,
+        AddCertificateError::InvalidCertificate(InvalidCertificateError::NonExistingAuthor { .. })
+    )
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn ok_non_root(env: &TestbedEnv) {
+    let env = env.customize(|builder| {
+        builder.new_user("bob");
+    });
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    let store = ops.store.for_write().await;
+    let (_, user_signed) = env.get_user_certificate("alice");
```

Comment:
rename that `alice_signed` `alice_dev1_signed` `bob_signed` (otherwise it's hard to understand the line `add_certificates_batch(&store, 0, [user_signed, device_signed, signed]`)

Reply:
and you should do the rename in the entire file (even if there is only a single certificate in a test, it's better to have it called `alice_signed` than just `signed`)

Reply:
Done